tests/resources/vrender.mjs

```diff
@@ -0,0 +1,906 @@
+/**
```

Comment:
This file is pretty mind blowing! This ain't everyday JS, this is shiny stuff. I find myself wondering how you did this!

I think, and correct me if I'm wrong, that this is a virtual DOM implementation to be used for rendering the tests, and I believe from our conversations that we wouldn't use it in the app, it would be confined to the aria-at repo. 

Which of course, the thought of having to come in here to fix something is making me sweat slightly, haha. It's frankly over my head right now. So my question is, is there a way to avoid having to maintain this ourselves? I'm sure the reasoning is solid but I was hoping you would take a little time to help me grasp it.

Finally, I was wondering if there's a good way to test this file specifically. How do I put it through its paces so I feel like I gave it the review it deserves?

Reply:
to echo @alflennik's comments, could you share your rationale for writing a custom implementation?

Any reason you didn't lean on a more commonly used dependency such as [lit-html](https://github.com/lit/lit/tree/main/packages/lit-html) or [snabbdom](https://github.com/snabbdom/snabbdom)?

Reply:
> This file is pretty mind blowing! This ain't everyday JS, this is shiny stuff. I find myself wondering how you did this!

> I think, and correct me if I'm wrong, that this is a virtual DOM implementation to be used for rendering the tests, and I believe from our conversations that we wouldn't use it in the app, it would be confined to the aria-at repo.

Yes. aria-at-app will use the javascript object produced by `instructionDocument`. aria-at itself renders that same object with this virtual DOM implementation.

> Which of course, the thought of having to come in here to fix something is making me sweat slightly, haha. It's frankly over my head right now. So my question is, is there a way to avoid having to maintain this ourselves? I'm sure the reasoning is solid but I was hoping you would take a little time to help me grasp it.
> 
> Finally, I was wondering if there's a good way to test this file specifically. How do I put it through its paces so I feel like I gave it the review it deserves?



> to echo @alflennik's comments, could you share your rationale for writing a custom implementation?
> 
> Any reason you didn't lean on a more commonly used dependency such as [lit-html](https://github.com/lit/lit/tree/main/packages/lit-html) or [snabbdom](https://github.com/snabbdom/snabbdom)?

1. Maintain behavior
    - Focus on the page header on first render
    - On keyboard event on a checkbox focus an adjacent checkbox
2. Keep the probability of adding bugs in this change
    - Produce the same html structure
3. Minimal external dependencies
4. Easy to turn instructionDocument into the current html shape
    - Many adjacent elements represent a semantic section without a parent element
    - Fragments like in React give an easy way to produce a html structure with semantic and not structural sections
    - Fragments help break up long sequences of adjacent elements

- LitHTML
    - Adds one dependency
    - Cannot produce the same dom
        - Inserts html comments at dynamic content locations
            - I find it hard this would have an impact but conservatively I want to avoid it
        - Supporting focus behavior
            1. Option: Use components and their update hook
                - Components are web components (custom elements)
                - This would put a change into the html structure
            2. Option: Manually find the right element after rendering without components and focus it

- Snabbdom
  - Adds a minimum of one dependency (components are a separate dependency)
  - Does not support fragments https://github.com/snabbdom/snabbdom/issues/560
    1. Option: change the html structure, add wrapping elements
    3. Option: Use less helper functions and repeat many elements more manually

- vrender
  - Internal dependency
  - Fragment support